pkg/destroy/destroyer.go

```diff
@@ -0,0 +1,78 @@
+package destroy
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	atd "github.com/openshift/hive/contrib/pkg/aws_tag_deprovision"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openshift/installer/pkg/asset/metadata"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// Destroyer allows multiple implementations of destroy
+// for different platforms.
+type Destroyer interface {
+	Run() error
+}
+
+// NewDestroyer returns Destroyer based on `metadata.json` in `rootDir`.
+func NewDestroyer(level log.Level, rootDir string) (Destroyer, error) {
+	raw, err := ioutil.ReadFile(filepath.Join(rootDir, metadata.MetadataFilename))
+	if err != nil {
+		return nil, err
+	}
+
+	var cmetadata types.ClusterMetadata
+	if err := json.Unmarshal(raw, &cmetadata); err != nil {
+		return nil, err
+	}
+
+	var ret Destroyer
+	switch {
+	case cmetadata.ClusterPlatformMetadata.AWS != nil:
+		ret = NewAWSDestroyer(level, &cmetadata)
+	case cmetadata.ClusterPlatformMetadata.Libvirt != nil:
+		// ret = NewLibvirtDestroyer(level, &cmetadata)
+		return nil, fmt.Errorf("libvirt destroyer is not yet supported")
+	default:
+		return nil, fmt.Errorf("couldn't find Destroyer for %q", metadata.MetadataFilename)
+	}
+	return ret, nil
+}
+
+// // NewLibvirtDestroyer returns libvirt Uninstaller from ClusterMetadata.
+// func NewLibvirtDestroyer(level log.Level, metadata *types.ClusterMetadata) *lpd.ClusterUninstaller {
+// 	return &lpd.ClusterUninstaller{
+// 		LibvirtURI: metadata.ClusterPlatformMetadata.Libvirt.URI,
+// 		Filter:     lpd.AlwaysTrueFilter(), //TODO: change to ClusterNamePrefixFilter when all resources are prefixed.
+// 		Logger: log.NewEntry(&log.Logger{
+// 			Out: os.Stdout,
+// 			Formatter: &log.TextFormatter{
+// 				FullTimestamp: true,
+// 			},
+// 			Hooks: make(log.LevelHooks),
+// 			Level: level,
+// 		}),
+// 	}
+// }
+
+// NewAWSDestroyer returns aws Uninstaller from ClusterMetadata.
+func NewAWSDestroyer(level log.Level, metadata *types.ClusterMetadata) *atd.ClusterUninstaller {
+	return &atd.ClusterUninstaller{
+		Filters: metadata.ClusterPlatformMetadata.AWS.Identifier,
+		Region:  metadata.ClusterPlatformMetadata.AWS.Region,
```

Comment:
What do you think about having optional CLI flags to bypass loading the metadata file, if not necessary? This would help us in Hive, but it's not urgent.

Reply:
@dgoodwin  there are plans to drop `clusterID` will be dropped from `InstallConfig`; 
currently the clusterID is used inside cluster as its identity => that will be moved to ClusterVersionOperator,
And the clusterID is also used to tag the AWS resources => this will be local identifier generated by installer and will be different from one used by CVO.
cc @crawford 

So `metadata.json` will become the only source where you'll be able to extract the identifier used but installer to tag aws resources.

Reply:
We were still hoping to use openshift-install to both provision and deprovision clusters in Hive (we should talk about moving the uninstall code to your repo at some point), but we're running in pods, there is no persistent disk state and thus no metadata.json. If there were a CLI flag or env var to just specify the cluster ID, we could use openshift-install without requiring any on disk state, and then we're all using the same uninstall path.

It's also a nice override for a user who lost their install dir and metadata, but can see the cluster UUID tag on their AWS objects.

NOTE: we do plan to extract the cluster UUID after provisioning and store in the status on the API object, so we would have this avail when  deprovisioning. Technically we could use that to generate a metadata.json and inject into the pod it's just more of a pain with Kube than just being able to specify command params or env vars.

Reply:
Discussed with @crawford on slack, he suggests just pushing the entire metadata file up as a configmap and mounting back in for deprovision, and initially this sounds like it should be ok. We'll go with this for now.